File: apps/api/src/services/market-data/dataset.ts
import type { MarketFeatureRow } from './features.js';
import type { PredictionTargetRow } from './target.js';

export interface TrainingDatasetRow {
  timestamp: string;
  features: Omit<MarketFeatureRow, 'timestamp'>;
  target: 0 | 1;
  futureTimestamp: string;
}

export function buildTrainingDataset(
  features: MarketFeatureRow[],
  targets: PredictionTargetRow[],
): TrainingDatasetRow[] {
  if (features.length === 0 || targets.length === 0) {
    return [];
  }

  const targetByTimestamp = new Map(
    targets.map((target) => [target.timestamp, target]),
  );

  const dataset: TrainingDatasetRow[] = [];

  for (const feature of features) {
    const target = targetByTimestamp.get(feature.timestamp);

    if (!target) {
      continue;
    }

    const featureValues = [
      feature.return_1,
      feature.log_return_1,
      feature.sma_10,
      feature.sma_20,
      feature.price_to_sma_10,
      feature.price_to_sma_20,
      feature.volatility_10,
      feature.volume_change_1,
      feature.rsi_14,
    ];

    if (featureValues.some((value) => value === null)) {
      continue;
    }

    dataset.push({
      timestamp: feature.timestamp,
      features: {
        close: feature.close,
        volume: feature.volume,
        return_1: feature.return_1,
        log_return_1: feature.log_return_1,
        sma_10: feature.sma_10,
        sma_20: feature.sma_20,
        price_to_sma_10: feature.price_to_sma_10,
        price_to_sma_20: feature.price_to_sma_20,
        volatility_10: feature.volatility_10,
        volume_change_1: feature.volume_change_1,
        rsi_14: feature.rsi_14,
      },
      target: target.target,
      futureTimestamp: target.futureTimestamp,
    });
  }

  return dataset.sort(
    (a, b) =>
      new Date(a.timestamp).getTime() -
      new Date(b.timestamp).getTime(),
  );
}

File: apps/ml-service/app/dataset.py
from dataclasses import dataclass
from typing import Any


FEATURE_NAMES = (
    "close",
    "volume",
    "return_1",
    "log_return_1",
    "sma_10",
    "sma_20",
    "price_to_sma_10",
    "price_to_sma_20",
    "volatility_10",
    "volume_change_1",
    "rsi_14",
)


@dataclass(frozen=True)
class ModelDataset:
    x: list[list[float]]
    y: list[int]
    timestamps: list[str]
    future_timestamps: list[str]


def training_rows_to_model_dataset(
    rows: list[dict[str, Any]],
) -> ModelDataset:
    """
    Convert API TrainingDatasetRow-shaped dictionaries into
    a numerical feature matrix and target vector.

    Feature order is defined by FEATURE_NAMES and must remain stable.
    """
    if not rows:
        raise ValueError("training rows cannot be empty")

    x: list[list[float]] = []
    y: list[int] = []
    timestamps: list[str] = []
    future_timestamps: list[str] = []

    for row_index, row in enumerate(rows):
        if not isinstance(row, dict):
            raise ValueError(f"row {row_index} must be an object")

        if "timestamp" not in row:
            raise ValueError(f"row {row_index} is missing timestamp")

        if "futureTimestamp" not in row:
            raise ValueError(f"row {row_index} is missing futureTimestamp")

        if "features" not in row:
            raise ValueError(f"row {row_index} is missing features")

        if "target" not in row:
            raise ValueError(f"row {row_index} is missing target")

        features = row["features"]

        if not isinstance(features, dict):
            raise ValueError(f"row {row_index} features must be an object")

        target = row["target"]

        if target not in (0, 1):
            raise ValueError(f"row {row_index} target must be 0 or 1")

        feature_values: list[float] = []

        for feature_name in FEATURE_NAMES:
            if feature_name not in features:
                raise ValueError(
                    f"row {row_index} is missing feature '{feature_name}'"
                )

            value = features[feature_name]

            if isinstance(value, bool) or not isinstance(value, (int, float)):
                raise ValueError(
                    f"row {row_index} feature '{feature_name}' must be numeric"
                )

            feature_values.append(float(value))

        x.append(feature_values)
        y.append(int(target))
        timestamps.append(str(row["timestamp"]))
        future_timestamps.append(str(row["futureTimestamp"]))

    return ModelDataset(
        x=x,
        y=y,
        timestamps=timestamps,
        future_timestamps=future_timestamps,
    )